src/tdbastar/tdbastar_epsilon.cpp

```diff
@@ -239,15 +248,41 @@ int lowLevelfocalHeuristicSequential(
           fcl::CollisionResult<double> result;
           fcl::collide(robot_objs[current_robot_idx], robot_objs[robot_idx],
                        request, result);
-          if (result.isCollision()) {
+          if (!result.isCollision() &&
+              heterogeneous) { // residual force should be checked if there is
+                               // no collision
+            check_rho = true;
+            auto dist = state1 - state2; // only pos, velocity
+            if (abs(dist(0)) < 0.2 && abs(dist(1)) < 0.2 &&
+                abs(dist(2)) < 1.5) {
+              float input[6] = {
+                  static_cast<float>(dist(0)), static_cast<float>(dist(1)),
+                  static_cast<float>(dist(2)), static_cast<float>(dist(3)),
+                  static_cast<float>(dist(4)), static_cast<float>(dist(5))};
+              nn_reset();
+              const auto nnType =
+                  (robot_types[robot_idx] == "integrator2_3d_large_v0")
+                      ? NN_ROBOT_LARGE
+                      : NN_ROBOT_SMALL;
+
+              nn_add_neighbor(input, nnType);           // self-robot type?
+              const float *rhoOutput = nn_eval(nnType); // in grams
+              rho += rhoOutput[0] / 1000 * 9.81;        // in Newtons
+            }
+          } else if (result.isCollision()) {
```

Comment:
This computation seems incorrect.
1.) Rho is not additive. You need to call nn_add_neighbor for each neighbor and eval only at the end.
2.) need to take into account own robot type *and* neighbor robot type.